### PR TITLE
Docs - performance guidelines

### LINKED ISSUE #4070: Docs: Explain metadata operations cost and affects better
Explain better how diffs, merges and commits work. Explain how metadata operations are linear to the size of the change. Explain common usage patterns like getting a change summary, running multiple commits or merges on a branch, etc. and how it might impact performance. Something like [this section of the S3 documentation](https://docs.aws.amazon.com/AmazonS3/latest/userguide/optimizing-performance-design-patterns.html) could be very helpful for users going to production

### PR DESCRIPTION
**Pull request by eden-ohana**
closes #4070

### CODE REVIEW

File: docs/understand/performance-design-patterns.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+## Import data to lakeFS
+When importing data to lakefs from external sources one time or regularly. Instead of copying it to the bucket you can use lakeFS import/ingest which will create a reference to the existing objects on your bucket and avoid the copy.
+
+## Access committed data
+Accessing committed data with the branch name: `lakefs://repo/main`
+will get the latest commit from the main branch. However, it will also try to fetch the uncommitted data from main and will perform more calls to the kv store.
+If needed to access only the committed data, use the commit id or tag id if exist: `lakefs://repo/commit-id-1`
+It will use only the commit layer and saves calls to the DB, allowing more fast access when listing.
+Uncommitted data is saved on the KV DB store and committed data on the object store, therefore it is preferred to work on committed data.
```
Comment (itaiad200) [inline docs/understand/performance-design-patterns.md:35]:
```suggestion
We usually read directly from a branch, e.g. `lakefs://repo/main/path`. A branch consist from a commit an uncommitted objects. lakeFS needs to look first for uncommitted changes for the path, and then for committed changes. If you're only interested in committed changes, you may try to read just the committed data by using `lakefs://repo/main@/path` and avoid unnecessary uncommitted changes lookup.
```

File: docs/understand/performance-design-patterns.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+## Direct upload
+Writing objects to the object store can be done using the API, lakectl, and the s3 gateway.
+Lakectl has the option to upload objects [directly](https://docs.lakefs.io/reference/commands.html#lakectl-fs-upload) to the object store (requires proper credentials).
+It is the faster option to upload an object with lakeFS, which will save latency from the gateway to s3.
+It will upload the data to the object store and then stage only the metadata to lakeFS.
```
Comment (itaiad200) [inline docs/understand/performance-design-patterns.md:41]: 1. Reading this I have the impression that I can only achieve that with lakectl. I can also do it with the API, or lakeFSFS, let mention and link them as well?
2. I would also mention the added benefit of data not going thru lakeFS server (scalability, etc.)

**Review by itaiad200 — CHANGES_REQUESTED** [review]
I think you really nail down the important use-cases, great work!
See my comments, most of them are minor rephrasing :)

File: docs/understand/performance-best-practices.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+Concurrent commits/merges on the same branch may result in a race and may need a retry if one fails.
+Therefore, it is suggested to avoid committing/merging on the same branch from multiple sources concurrently.
+
+## Use zero-copy import
+To import object into lakeFS, either a single time or regularly, lakeFS offers a [zero-copy import](https://docs.lakefs.io/setup/import.html#zero-copy-import) feature.
+Use this feature to import a large number of objects to lakeFS, instead of simply copying them into your repository.
+This feature will create a reference to the existing objects on your bucket and avoid the copy.
+
+## Read data using the commit ID
+In cases where you are only interested in reading committed data, use a commit ID (or a tag ID) in your path (e.g: `lakefs://repo/a1b2c3`).
```
Comment (itaiad200) [inline docs/understand/performance-best-practices.md:32]: Please also include the special notation we used for reading committed data from a branch. See my previous comment.

File: docs/understand/performance-best-practices.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+When accessing data using the branch name (e.g. `lakefs://repo/main`) lakeFS will also try to fetch uncommitted data, which may result in reduced performance.
+For more information, see [how uncommitted data is managed in lakeFS](https://docs.lakefs.io/understand/versioning-internals.html#representing-references-and-uncommitted-metadata)
+
+## Operate directly on the storage
+Sometimes, storage operations can become a bottleneck. For example, when your data pipelines uploads very big objects.
```
Comment (itaiad200) [inline docs/understand/performance-best-practices.md:37]:
```suggestion
Sometimes, storage operations can become a bottleneck. For example, when your data pipelines upload many big objects.
```

**Review by itaiad200 — CHANGES_REQUESTED** [review]
Getting closer :) see my comments.

File: docs/understand/performance-best-practices.md (hunk 4 above)
Comment (eden-ohana) [reply docs/understand/performance-best-practices.md:37]: done

File: docs/understand/performance-best-practices.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+As a rule of thumb, try to keep your commits smaller than 1M objects.
+
+## Avoid concurrent commits/merges
+Concurrent commits/merges on the same branch may result in a race and may need a retry if one fails.
+Therefore, it is suggested to avoid committing/merging on the same branch from multiple sources concurrently.
+
+## Use zero-copy import
+To import object into lakeFS, either a single time or regularly, lakeFS offers a [zero-copy import](https://docs.lakefs.io/setup/import.html#zero-copy-import) feature.
+Use this feature to import a large number of objects to lakeFS, instead of simply copying them into your repository.
+This feature will create a reference to the existing objects on your bucket and avoid the copy.
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply docs/understand/performance-best-practices.md:29]: done

File: docs/understand/performance-best-practices.md [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+## Overview
+Use this guide to achieve the best performance with lakeFS.
+
+## Avoid huge commits
+If you're having issues with slow commits, consider performing smaller commits. The more objects in the commit, the longer the commit will take to complete.
+Moreover, a commit should represent a meaningful point in your data's lifecycle, so limiting the commit size will create a more comprehensible commit history.
+As a rule of thumb, try to keep your commits smaller than 1M objects.
+
+## Avoid concurrent commits/merges
+Concurrent commits/merges on the same branch may result in a race and may need a retry if one fails.
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply docs/understand/performance-best-practices.md:23]: done

File: docs/understand/performance-best-practices.md (hunk 3 above)
Comment (eden-ohana) [reply docs/understand/performance-best-practices.md:32]: I checked and we don't support @ anymore

Comment (itaiad200) [reply docs/understand/performance-best-practices.md:32]: https://github.com/treeverse/lakeFS/blob/08a3e8798c3616b4d1bbef3d675d2df9ac27f56e/pkg/graveler/ref/resolve_ref.go#L55 ?

**Review by itaiad200 — APPROVED** [review]
See my single comment about `@` for committed entries. Thanks!

**Review by itaiad200 — APPROVED** [review]

**Pull request merged**